src/backend/distributed/utils/resource_lock.c

```diff
@@ -342,3 +342,42 @@ LockMetadataSnapshot(LOCKMODE lockMode)
 
 	(void) LockRelationOid(DistNodeRelationId(), lockMode);
 }
+
+
+/*
+ * LockNodeId returns after grabbing a lock for distribution metadata related
+ * to the specified shard, blocking if required. Any locks acquired using this
+ * method are released at transaction end.
+ */
+void
+LockNodeId(int64 nodeId, LOCKMODE lockMode)
```

Comment:
I don't like that we have to add a whole new lock tag just for this one pathway. I believe you could get away with simply locking the row in `pg_dist_node`. In particular, you error out if the node can't be found, so you only proceed to locking if the row exists. This should have identical semantics to something like `SELECT ... FOR UPDATE`, so could probably be implemented by simply finding the row and retaining an exclusive lock on it to the end of the operation. If the row isn't found, error out (see the call site to `TryLockNodeId`, etc.)